### PR TITLE
ClassDictionary value type any

### PR DESCRIPTION
**Pull request by jeffijoe**
@vvakame as requested - #11317

### CODE REVIEW

File: classnames/classnames.d.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@
 declare type ClassValue = string | number | ClassDictionary | ClassArray;
 
 interface ClassDictionary {
-	[id: string]: boolean;
+	[id: string]: any;
```
Comment (adidahiya) [inline classnames/classnames.d.ts]: no, I don't think this should be expanded to `any`.

what problem is this really addressing? 

your example in #11317 seems to work already: https://www.typescriptlang.org/play/#src=declare%20type%20ClassValue%20%3D%20string%20%7C%20number%20%7C%20ClassDictionary%20%7C%20ClassArray%3B%0D%0A%0D%0Ainterface%20ClassDictionary%20%7B%0D%0A%09%5Bid%3A%20string%5D%3A%20any%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20ClassArray%20extends%20Array%3CClassValue%3E%20%7B%20%7D%0D%0A%0D%0Ainterface%20ClassNamesFn%20%7B%0D%0A%09(...classes%3A%20ClassValue%5B%5D)%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Adeclare%20var%20classNames%3A%20ClassNamesFn%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0Ainterface%20MyProps%20%7B%0D%0A%09foo%3F%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Aconst%20myProps%3A%20MyProps%20%3D%20%7B%0D%0A%09%2F%2F%20nothing%0D%0A%7D%3B%0D%0A%0D%0Aconst%20classes%20%3D%20classNames(%7B%0D%0A%09%22some-class%22%3A%20myProps.foo%2C%0D%0A%7D)%0D%0A

**Review by adidahiya — CHANGES_REQUESTED** [review]

File: classnames/classnames.d.ts (hunk 1 above)
Comment (jeffijoe) [inline classnames/classnames.d.ts]: @adidahiya check #11317

Comment (jeffijoe) [inline classnames/classnames.d.ts]: @adidahiya try pasting this:

```ts
declare type ClassValue = string | number | ClassDictionary | ClassArray;

interface ClassDictionary {
	[id: string]: boolean;
}

interface ClassArray extends Array<ClassValue> { }

interface ClassNamesFn {
	(...classes: ClassValue[]): string;
}

declare var classNames: ClassNamesFn;


interface MyProps {
	foo?: string;
}

const myProps: MyProps = {
	// nothing
};

const classes = classNames({
	"some-class": myProps.foo,
})
```

Comment (adidahiya) [inline classnames/classnames.d.ts]: whoops, my bad there. [here's an updated playground](https://www.typescriptlang.org/play/#src=declare%20type%20ClassValue%20%3D%20string%20%7C%20number%20%7C%20ClassDictionary%20%7C%20ClassArray%3B%0D%0A%0D%0Ainterface%20ClassDictionary%20%7B%0D%0A%09%5Bid%3A%20string%5D%3A%20any%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20ClassArray%20extends%20Array%3CClassValue%3E%20%7B%20%7D%0D%0A%0D%0Ainterface%20ClassNamesFn%20%7B%0D%0A%09(...classes%3A%20ClassValue%5B%5D)%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Adeclare%20var%20classNames%3A%20ClassNamesFn%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0Ainterface%20MyProps%20%7B%0D%0A%09foo%3F%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Aconst%20myProps%3A%20MyProps%20%3D%20%7B%0D%0A%09%2F%2F%20nothing%0D%0A%7D%3B%0D%0A%0D%0Aconst%20classes%20%3D%20classNames(%7B%0D%0A%09%22some-class%22%3A%20myProps.foo%2C%0D%0A%7D)%0D%0A#src=declare%20type%20ClassValue%20%3D%20string%20%7C%20number%20%7C%20ClassDictionary%20%7C%20ClassArray%3B%0D%0A%0D%0Ainterface%20ClassDictionary%20%7B%0D%0A%09%5Bid%3A%20string%5D%3A%20boolean%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20ClassArray%20extends%20Array%3CClassValue%3E%20%7B%20%7D%0D%0A%0D%0Ainterface%20ClassNamesFn%20%7B%0D%0A%09(...classes%3A%20ClassValue%5B%5D)%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Adeclare%20var%20classNames%3A%20ClassNamesFn%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0Ainterface%20MyProps%20%7B%0D%0A%09foo%3F%3A%20boolean%3B%0D%0A%7D%0D%0A%0D%0Aconst%20myProps%3A%20MyProps%20%3D%20%7B%0D%0A%09%2F%2F%20nothing%0D%0A%7D%3B%0D%0A%0D%0Aconst%20classes%20%3D%20classNames(%7B%0D%0A%09%22some-class%22%3A%20myProps.foo%2C%0D%0A%7D)%0D%0A) that matches #11317

I'm still reluctant to introduce this change in general; I would prefer to be stricter about the types and not implicitly coerce values to booleans.

Comment (jeffijoe) [inline classnames/classnames.d.ts]: This is what I am seeing:

image

In that case I'm gonna have to use my own typings.

Comment (adidahiya) [inline classnames/classnames.d.ts]: sorry, I am failing at sharing from the playground today :)

screencap:

image

[link to a working playground](https://www.typescriptlang.org/play/index.html#src=declare%20type%20ClassValue%20%3D%20string%20%7C%20number%20%7C%20ClassDictionary%20%7C%20ClassArray%3B%0D%0A%0D%0Ainterface%20ClassDictionary%20%7B%0D%0A%09%5Bid%3A%20string%5D%3A%20boolean%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20ClassArray%20extends%20Array%3CClassValue%3E%20%7B%20%7D%0D%0A%0D%0Ainterface%20ClassNamesFn%20%7B%0D%0A%09(...classes%3A%20ClassValue%5B%5D)%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Adeclare%20var%20classNames%3A%20ClassNamesFn%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0Ainterface%20MyProps%20%7B%0D%0A%09foo%3F%3A%20boolean%3B%0D%0A%7D%0D%0A%0D%0Aconst%20myProps%3A%20MyProps%20%3D%20%7B%0D%0A%09%2F%2F%20nothing%0D%0A%7D%3B%0D%0A%0D%0Aconst%20classes%20%3D%20classNames(%7B%0D%0A%09%22some-class%22%3A%20myProps.foo%2C%0D%0A%7D)%0D%0A)

Comment (jeffijoe) [inline classnames/classnames.d.ts]: Sorry, the reason it's not working for me is because I am using TS strict-mode :)

**Comment by mhegazy** [comment]
PR rejected by reviewer. closing.

**Comment by jeffijoe** [comment]
Too bad, it's not very compatible with TS2.0 strict mode.

**Comment by mhegazy** [comment]
what is "TS2.0 strict mode"?

**Comment by jeffijoe** [comment]
Sorry, strictNullChecks I believe

~ Jeff Hansen

On 3 Oct 2016, at 21.59, Mohamed Hegazy <notifications@github.com<mailto:notifications@github.com>> wrote:


what is "TS2.0 strict mode"?

-

**Comment by mhegazy** [comment]
so then this should `boolean | undefined` and not `any`.

**Comment by jeffijoe** [comment]
That was one of my suggestions in the original issue as well.

~ Jeff Hansen

On 3 Oct 2016, at 22.49, Mohamed Hegazy <notifications@github.com<mailto:notifications@github.com>> wrote:


so then this should boolean | undefined and not any.

-

**Comment by mhegazy** [comment]
i think changing it to `boolean | undefined` should be fine. but please submit this PR against https://github.com/DefinitelyTyped/DefinitelyTyped/tree/types-2.0

**Comment by jeffijoe** [comment]
Whats different in that branch?

~ Jeff Hansen

On 3 Oct 2016, at 23.02, Mohamed Hegazy <notifications@github.com<mailto:notifications@github.com>> wrote:


i think changing it to boolean | undefined should be fine. but please submit this PR against https://github.com/DefinitelyTyped/DefinitelyTyped/tree/types-2.0

-

**Comment by mhegazy** [comment]
master is still TS 1.8 branch, and the `undefined` type is only supported in TS 2.0. branch types-2.0 is the TS 2.0 branch. in the 4 or 5 weeks we will switch master to be a TS 2.0 branch. but until then TS-2.0-specific changes should go to that branch.

**Comment by jeffijoe** [comment]
Got it

~ Jeff Hansen

On 3 Oct 2016, at 23.11, Mohamed Hegazy <notifications@github.com<mailto:notifications@github.com>> wrote:


master is still TS 1.8 branch, and the undefined type is only supported in TS 2.0. branch types-2.0 is the TS 2.0 branch. in the 4 or 5 weeks we will switch master to be a TS 2.0 branch. but until then TS-2.0-specific changes should go to that branch.

-

**Comment by jeffijoe** [comment]
@mhegazy it appears to have already been implemented in `types-2.0` - https://github.com/DefinitelyTyped/DefinitelyTyped/pull/11584/commits/973faa0dbd2feb7b208643cfc725094298611333

**Comment by mhegazy** [comment]
sweet. thanks!

**Pull request closed without merging**